### DISCUSSION TITLE
object detection (dry cat food) (Custom Models ?)

### DISCUSSION THREAD

**Discussion by euDominic** [blakeblackshear/frigate#4793] [General]
However, I have a special use case for object detection.
 
I have two cats and each has its own automated food bowl. 
Now, I would only give the cats food when it is certain that the bowls are empty.
 
The object detection correctly registers that bowls are visible, but not whether they are with content or not.
 
Now the question, is it possible from your side to create a model for the cat food (dry)? Or do you have an idea how I can otherwise implement the recognition?
A custom model would probably be the best option, but I lack the knowledge and experience to realize it myself.

Maybe someone can help me.
![snapshot-1671835420 483655-nugefq](https://user-images.githubusercontent.com/88461637/209438375-1e06340e-bda6-458c-bc79-29795e3a503b.jpg)
![snapshot-1671835808 055596-exstyz](https://user-images.githubusercontent.com/88461637/209438376-f46ea867-59e1-4a8c-b896-f891a516e085.jpg)
![snapshot-1671883841 958702-133i8b](https://user-images.githubusercontent.com/88461637/209438377-273d1843-866c-477a-a18c-c49e605b70f1.jpg)
![snapshot-1671883365 884852-dqcom3](https://user-images.githubusercontent.com/88461637/209438378-133e2414-f8ea-4e81-8fc2-ef6131d56124.jpg)
![snapshot-1671883387 336785-jwqonp](https://user-images.githubusercontent.com/88461637/209438379-06184e58-a5d0-492a-9645-ea299df9f841.jpg)
![snapshot-1671883819 675573-s89tne](https://user-images.githubusercontent.com/88461637/209438380-cb3c4b27-322c-4898-89b5-d84a43f2f122.jpg)

**Comment by klutzzykitty** [comment]
@euDominic 
You can create a custom model with https://teachablemachine.withgoogle.com/ Teachable machine for simple tasks. 
The same can be converted to EdgeTPU model for deployment.

**Reply by Gie92** [reply]
Hi! @klutzzykitty 

Just wondering if you were able to get the custom models from teachable machine to work with frigate?

**Reply by jeeftor** [reply]
Mee tooo!

**Reply by heisenberg2980** [reply]
I tried and I think I got pretty close, but after adding the model to my frigate I am getting an error.

This is what I added to my config (after uploading the custom model generated by teachable machine, and mapping the /custom_model folder so it is visible inside the container):

```
detectors:
  coral:
    type: edgetpu
    device: usb
    # Optional: model config
#    model:
#      path: /custom_model/model_edgetpu.tflite

model:
  path: /custom_model/model_edgetpu.tflite
  labelmap_path: /custom_model/labels.txt
  width: 224
  height: 224
  input_tensor: "nhwc"
  input_pixel_format: "bgr"
```

And this is the error I am getting:

```
2023-08-08 19:43:10.057371923  Process detector:coral:
2023-08-08 19:43:10.057856060  Traceback (most recent call last):
2023-08-08 19:43:10.057857362    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2023-08-08 19:43:10.057858084      self.run()
2023-08-08 19:43:10.057858615    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2023-08-08 19:43:10.057862472      self._target(*self._args, **self._kwargs)
2023-08-08 19:43:10.057863033    File "/opt/frigate/frigate/object_detection.py", line 121, in run_detector
2023-08-08 19:43:10.057885625      detections = object_detector.detect_raw(input_frame)
2023-08-08 19:43:10.057886486    File "/opt/frigate/frigate/object_detection.py", line 71, in detect_raw
2023-08-08 19:43:10.057886887      return self.detect_api.detect_raw(tensor_input=tensor_input)
2023-08-08 19:43:10.057887298    File "/opt/frigate/frigate/detectors/plugins/edgetpu_tfl.py", line 59, in detect_raw
2023-08-08 19:43:10.057903127      class_ids = self.interpreter.tensor(self.tensor_output_details[1]["index"])()[0]
2023-08-08 19:43:10.057912935  IndexError: list index out of range
```

**Reply by Gie92** [reply]
Teachable Machine has the correct uint8 input, but needs the correct outputs, which is (4X) float32 outputs, the closest model maker I found is edge impulse, but the input is int8... And I can't find a way to convert it to uint8...

**Reply by klutzzykitty** [reply]
> Hi! @klutzzykitty
> 
> Just wondering if you were able to get the custom models from teachable machine to work with frigate?

Hey  @Gie92  @jeeftor  @Heisenberg2980  ! Was offline for a while . 

Has anyone tried exporting the model to Tflite format from teachable machine and then using edgetpu compiler to manually convert the downloaded model to edgetpu compatible format. 

Edge tpu supports both uint8 and fp32 optimized models. 
Ex reference: `pipeline.train_config.fine_tune_checkpoint = '/content/pretrained_model/ssdlite_mobiledet_edgetpu_320x320_coco_2020_05_19/fp32/model.ckpt'
`

`pipeline.train_config.fine_tune_checkpoint = '/content/pretrained_model/ssdlite_mobiledet_edgetpu_320x320_coco_2020_05_19/uint8/model.ckpt'
`

**Reply by Gie92** [reply]
No worries, yeah but unfortunately the outputs don't match...

**Reply by klutzzykitty** [reply]
I might be missing something here. 
Is there a different output format required for edgetpu compiler to convert an input TFlite model? 

Also, can you share where the 4x float32 information comes from? 

Teachable machine is a coral TPU project by itself, I might be missing some background on whats actually going wrong.

**Comment by Equidamoid** [comment]
Using NN for such a task feels like an overkill. How about something dumber? Count pixels of specific color/brightness within the boundaries of the bowls? Getting white (ceramic/enameled) bowls will help a lot with that. 
Not sure how hard it is to add it to Frigate (and mandatory Docker usage won't help with that), but you can always just make a separate script that just fetches a frame every x min, does the processing and shoots the result in whatever direction you want.

For a crazier option: put a load cell under the bowl (I actually want to weigh the whole feeder, to know when it's time to refill it). 

And a radical alternative: measure how much food the feeders spit out (approximately), and give them the recommended amount/day.

**Comment by heisenberg2980** [comment]
@euDominic I think Swatch is what you are looking for: https://github.com/NickM-27/swatch
